### PR TITLE
Introducing the kubernetes modules

### PR DESCRIPTION
**Pull request by flavio**
### What does this PR do?

This PR leads the foundation to manage Kubernetes resources right from salt.

Currently it's possible to manage Kubernetes [deployments](https://kubernetes.io/docs/user-guide/deployments/) and [services](https://kubernetes.io/docs/user-guide/services/).

This allows for example to deploy a relatively complex kubernetes 101 example like the [guestbook](https://github.com/kubernetes/kubernetes/tree/master/examples/guestbook) one.

The interaction with kubernetes is achieved using the [official python library for kubernetes](https://github.com/kubernetes-incubator/client-python).

### What issues does this PR fix or reference?

This is an idea @dmacvicar and I had at @SUSE. You can find more about it inside of this [RFC](https://gist.github.com/dmacvicar/1793395c2284e23e9cac8d210d239dab).

### Tests written?

No.

### Demo

I created a demo repository containing all the resources needed to play around with this code. It also includes some asciicast showing the integration.

You can find it [here](https://github.com/flavio/salt-kubernetes-demo).

### RFC

This PR is a work in progress. I created that to gather feedback from more experienced Salt users and developers. Actually this is the first time I extend Salt 😅 

These are the open questions I have:

  * What do you think about the way to declare kubernetes resources inside of `.sls` files? I wanted to allow users to simply copy & paste kubernetes manifests into the `.sls` files. At the same time I didn't want to specify all the possible attributes of all the resources. That's why I decided to take just the resource `meta` and `spec`.
  * I don't like how the code deals when a service/deployment is already present: it just replaces it. The code should be changed to compare the actual object with the one specified by the user, find the differences and then fix them. This would provide more meaningful diff reports.
  * Right now the kubernetes state files have to be assigned against a minion. However this does not make sense because the interaction with the kubernetes cluster could be done even by the salt-master node. It's just a matter of installing the kubernetes-python library on the salt-master and then put the address of the kubernetes API server into the pillar. Is there a smart way to achieve that?
  * I wanted to make the code as generic as possible when dealing with kubernetes objects. However I think we will have to end up dealing with specific code for each object. Isn't there the risk of having huge `kubernetes.py` files?

### CODE REVIEW

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: We'd like to use the version of `six` that we bundle. This could be `import salt.ext.six as six`.

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: We prefer not to have bare exceptions. We prefer `except Exception` here, which should catch all cases.

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: We prefer not to print from execution modules. Instead, either use `log.exception(exc)` and we'll format and log the exception for you.

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: Same comment about `print` here.

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: Another print that needs to be `log.exception`

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: log.exception

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: log.exception

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: log.exception

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: log.exception

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: log.exception

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: We prefer single-quotes for docstrings, per our style-guide

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: single-quotes

File: salt/modules/kubernetes.py
Comment (cachedout) [inline salt/modules/kubernetes.py]: single-quotes

File: salt/states/kubernetes.py
Comment (cachedout) [inline salt/states/kubernetes.py]: single-quotes on these arsg, please

**Review by cachedout — COMMENTED** [review]
Mostly some comments here about changing things to confirm to the saltstack-style guide. I've also asked others to look at this and we'll respond to your broader questions as well.

File: salt/modules/kubernetes.py
Comment (s0undt3ch) [inline salt/modules/kubernetes.py]: Preferably, catch only specific exception that might occur instead of all.

Comment (flavio) [inline salt/modules/kubernetes.py]: Multiple types of exception can be raised when trying to connect to the API server, ranging from the server being offline to the credentials entered by the users being incorrect.

If you want I can try to figure out all the types of exceptions returned, but I fear something might be omitted...

In the meantime I changed the code to catch `Exception`.

**Comment by flavio** [comment]
Sorry about `"` vs `'`, I fixed that but I forgot to update my PR...

BTW, I also tried to run the salt linter but it failed with some weird error. I just ran the code against vanilla pylint and pyflakes

**Comment by s0undt3ch** [comment]
You need to install `saltpylint` and use `.testing.pylintrc` the you should have lint results close to ours.

**Comment by s0undt3ch** [comment]
*then you should....

**Comment by flavio** [comment]
@s0undt3ch: I fixed my issue, I had installed `saltpylint` using pip3 instead of pip2...

I fixed the relevant issues, I can sort out the warnings later. I expect more feedback to come and hence to change some implementation details.

BTW, I just realized there was already some code to interact with kubernetes. 🤦‍♂️  Sorry, we can merge the two codebases together of course.

**Comment by cachedout** [comment]
@flavio Yes, you beat me to that question. ;]

I'd like to see if @titilambert, who wrote the original k8 support would like to take a look here and perhaps work with us on a path forward for both the old module and the new.

**Comment by cachedout** [comment]
I'd like to bring this PR to the attention of @terminalmage @gtmanfred and @rallytime for ideas about how to merge this with existing functionality.

cc: @thatch45

**Review by thatch45 — APPROVED** [review]
First and foremost, I want to get this merged in!
As for the k8 module, I would say that we should look at converging, I don't care which name, but that we should look at this from a functionality perspective and migrate functionality from k8 into this module or vis-versa. Since we can do module aliasing better now we should be able to sketch out a convergence plan, but yes, for now, lets get this in!

**Comment by terminalmage** [comment]
The existing functionality (``salt/{states,modules}/k8s.py``) seem less comprehensive than this, so my first inclination is to favor the new functionality over the existing. Also, the architecture of the k8s.py execution module does not follow our conventions. The function names seem to mostly mirror the state functions, and they also return highstate return data structures. The more "salty" way to do states and execution modules is to have the execution module provide a more general interface to the thing being managed. This includes having functions to gather information, and functions to make changes. The idea is for the execution module to be useful both to states and also useful in their own right on the CLI. The existing k8s state module offloads _all_ of the work to the execution module, which is less than optimal because it makes the functions in the execution module less flexible. State module functions should leverage the more general interface provided in execution modules, and should themselves perform the logic which determines whether or not something needs to be done. A pseudocode approach for a good state module function would be something like:

```
perform checks on input to detect conflicting/mutually-exclusive arguments, and exit with a false result if any are detected

use a remote execution function to gather information about the thing being managed

if no changes need to be made:
    exit with a True result and a comment noting that thing X is already configured as specified

if __opts__['test']:
    exit with a None result and a comment noting that thing X would change, and describe in what way

make changes

re-perform the earlier check, using a remote execution function to gather information about the thing being managed

check again if changes need to be made

if changes need to be made:
    exit with a False result and a comment noting that thing X failed to be managed as expected, including any changes made in the ``changes`` key in the state return

exit with a True result and a comment noting that thing X was successfully updated, including any changes made in the ``changes`` key in the state return
```

We should check k8s.py for any functionality missing from the new code submitted by @flavio and add it to the new module. Both modules should then be updated with information in their docstrings notifying users that there are two modules, and that the new one will be preferred moving forward. A timetable for deprecation of the old module should also be in the docstring.

I have recent experience with doing this with our docker support. We kept the old code around for a few releases (with ample warning in the documentation), before recently moving them to [salt-contrib](https://github.com/saltstack/salt-contrib) in develop. The old state/execution module can be moved there once it's been deprecated long enough to be removed. Our general rule of thumb has been to allow at least 2 feature release cycles for something to be removed once deprecated (for docker we waited 3, I think, but part of that was a lack of spare cycles to devote to it).

Another way to encourage people to move to newer kubernetes states would be to add a warning to the state returns in k8s.py (see [here](https://github.com/saltstack/salt/blob/5ad4769/salt/states/docker.py#L131-L137) for an example). Basically, adding a ``warnings`` key to the state return (which should be a list of strings) will result in those warnings being printed to the console in the highstate return. The usage of ``salt.utils.warn_until()`` adds a DeprecationWarning to the log file.

Once the old module has been removed after the deprecation period is through, we can use a feature I recently added to make the ``k8s`` state/module names reference the new code. By using ``__virtual_aliases__`` (example [here](https://github.com/saltstack/salt/blob/994892b/salt/modules/docker.py#L347)) kubernetes.py would import both as ``kubernetes`` and as ``k8s``.

**Review by terminalmage — CHANGES_REQUESTED** [review]
See the comment I left earlier for my review,

**Comment by damon-atkins** [comment]
@terminalmage    please add your comments to the contribution guide.

**Comment by terminalmage** [comment]
@damon-atkins I have plenty of high-priority issues to keep me busy at the moment, so while it would be nice to do so, it's not a priority for me at this time.

**Comment by terminalmage** [comment]
However, if you'd really like to see this information there, we do accept pull requests.

**Comment by cachedout** [comment]
@flavio I would like to get your feedback on the earlier comments about integration from @terminalmage

**Comment by flavio** [comment]
Sorry for the late response, I've been pretty busy lately. The good news is: next week I'll have plenty of time to hack on this PR.

I can definitely take into account what @terminalmage wrote.

However, please provide some feedback about the "request for comments" points I raised inside of the body of the PR. I wouldn't like to implement the other missing features in a wrong way.

**Comment by terminalmage** [comment]
> What do you think about the way to declare kubernetes resources inside of .sls files? I wanted to allow users to simply copy & paste kubernetes manifests into the .sls files. At the same time I didn't want to specify all the possible attributes of all the resources. That's why I decided to take just the resource meta and spec.

I am not familiar with any of this, so I will defer to someone with knowledge of and experience with k8s.

> I don't like how the code deals when a service/deployment is already present: it just replaces it. The code should be changed to compare the actual object with the one specified by the user, find the differences and then fix them. This would provide more meaningful diff reports.

Yes, definitely. My pseudocode example sort of outlined an approach for this. The convention in Salt is to have an information gathering function which you can compare the state input to in some way to see if changes need to be made. This is how we achieve idempotence. A good example of this is the ``pkg.installed`` state. It runs ``pkg.list_pkgs``, which gives it a dictionary package names to installed versions. We can then check the state input and see if the desired packages (and optionally versions) match what is already installed. Anything that does *not* match becomes a candidate for installation. If we do not find that any changes need to be made, we can return early with an empty ``changes`` dict in the return data, and a ``True`` result.

> Right now the kubernetes state files have to be assigned against a minion. However this does not make sense because the interaction with the kubernetes cluster could be done even by the salt-master node. It's just a matter of installing the kubernetes-python library on the salt-master and then put the address of the kubernetes API server into the pillar. Is there a smart way to achieve that?

So, states have to be run by minions. Salt does have a method for running code that is purely master-side, we call these "runners" and you can check ``salt/runners/`` in the source code for examples.

What you _could_ do is have a state and execution module, and also a complimentary runner module. The runner module could run the same functions as the state module. You may even be able to abstract a great deal of the logic into a salt/utils/k8s.py which both the state and runner draw from. That way you could support both methods of managing kubernetes and not tie people to one or the other. We have a class called ``salt.minion.MasterMinion`` which you'll see referenced in some of our existing runner modules, which allows the master to load the execution modules as if it were a minion, which would give it access to the functions in ``salt/modules/k8s.py``.

> I wanted to make the code as generic as possible when dealing with kubernetes objects. However I think we will have to end up dealing with specific code for each object. Isn't there the risk of having huge kubernetes.py files?

Again, I lack the knowledge and experience necessary to be helpful with this question. Sorry!

**Comment by cachedout** [comment]
I think we've had a good discussion here and I'd like us to figure out the path here. We'll want to get this in fairly soon. The first thing we need to do here is to change the base branch to `Nitrogen` so that we get it in the release. @flavio do you want to handle that?

Second, let's agree if there are any changes that need to be made here or whether we can merge this as it sits. So, final call for comments on this. :]

**Comment by flavio** [comment]
> I think we've had a good discussion here and I'd like us to figure out the path here. We'll want to get this in fairly soon. The first thing we need to do here is to change the base branch to Nitrogen so that we get it in the release. @flavio do you want to handle that?

Sure, that would be my pleasure. I'll probably need some instructions given this is the 1st time I do something like that against Salt :)

> Second, let's agree if there are any changes that need to be made here or whether we can merge this as it sits. So, final call for comments on this. :]

/me waits

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
@flavio I have updated the base branch to Nitrogen.

**Review by gtmanfred — APPROVED** [review]

**Comment by rallytime** [comment]
I have switched the base branch on this pull request from `nitrogen` to `2017.7` in preparation for moving from a code-named release to a version number.

**Comment by flavio** [comment]
Thanks, in the meantime I've been working to bring feature parity with the existing `k8s` module/states.
I'm almost done...

**Comment by cachedout** [comment]
@flavio Can you give us an ETA on this? We want to tag the release very soon.

**Comment by cachedout** [comment]
@flavio Please contact me ASAP. I am going to merge this as-is if I don't hear from you so that we get it in before the release. Thanks.

**Comment by flavio** [comment]
@cachedout I'm currently on the `#salt` channel on freenode (nick `flavio`)

**Comment by flavio** [comment]
I finished to implement all the missing features I mentioned to @cachedout on irc.

Unfortunately kubernetes' label management is not implemented by this new module due to [this open bug](https://github.com/kubernetes-incubator/client-python/issues/240) inside of the kubernetes client.

Kubernetes' labels are the only resource managed by `k8s` state module. Given the new code doesn't manage them I didn't introduce any warning message inside of the `k8s` states. What do you think about that, should I still do that? If that's the case which version is going to be the last one supporting the old `k8s` module?

**Comment by cachedout** [comment]
@flavio I think that a warning would be a good idea if you can get to it quickly. Otherwise we'll merge this as-is so we can make the deadline. Thanks!

**Comment by flavio** [comment]
@cachedout sure, I can add a warning.

I'll add something like @terminalmage suggested (see [this](https://github.com/saltstack/salt/blob/5ad4769/salt/states/docker.py#L131-L137). Which version of salt should I specify, here it mentions `Fluorine`).

**Comment by cachedout** [comment]
@flavio Fluorine would be appropriate, I think.

**Comment by flavio** [comment]
Great news, using tcpdump I've been able to handle also labels. I'm currently working on that and really close to get a fix.

I should have everything in place (labels + deprecation warning) by 22:00 CEST which should translate to [14:00 MST](https://www.timeanddate.com/worldclock/meetingtime.html?day=21&month=6&year=2017&p1=5521&p2=75&iv=0) (dinner with my daughters is involved 😄).

Does it work for you?

**Comment by cachedout** [comment]
@flavio Yes that works for us. Thank you!

**Comment by cachedout** [comment]
@flavio Before you finish this up could you please also rebase it against the head of the branch? Thanks.

**Comment by flavio** [comment]
@cachedout I'm done 🎉 

Please take a look at the deprecation message I wrote and tell me if I've to change it somehow.

**Comment by flavio** [comment]
... and done also with the rebase against `2017.7`

**Comment by flavio** [comment]
@cachedout: now I'm really done, we now have complete feature parity between the `k8s` and `kubernetes` modules.

It's getting late here, if you need something drop me a mail and I'll do my best to fix the code. Otherwise I think we are good to go.

**Comment by cachedout** [comment]
Merged! Thanks for everything, @flavio. This is really great!

File: salt/states/kubernetes.py
Comment (isbm) [inline salt/states/kubernetes.py]: @flavio why not just:
```python
if not metadata or not spec:
    return {}
```

**Comment by isbm** [comment]
@flavio 
> ## Tests written?
> No.

why-meme-4-maker

**Comment by cachedout** [comment]
@isbm The majority of this code is just wrappers around a third-party library. What functionality, specifically, do you think should be tested here and how do you propose that it's done?

**Comment by isbm** [comment]
@cachedout yep, wrappers! And this is _exactly_ why I would be overparanoid making even more tests! Just like we did a wrapper around Zypper, which differs from version to version of the platform and you never know what you are going to hit the next call on another update. So the output of the external program is mocked, the rest of the logic should be tested at least for sake of regressions. And this module fairly easy to test, would take a day, basically to implement all necessary UTs.

**Comment by flavio** [comment]
I always wanted to provide test coverage of the code I wrote. I would  love to come up with a set of integration tests that would target different releases of kubernetes.

A dream setup would be something like that:

  1. Start a a specific version of kubernetes using [minikube](https://github.com/kubernetes/minikube)
  1. Run some real world operations against the kubernetes cluster, something like the examples I provided inside of the demo project I linked at the top of the PR.
  1. Collect the results

Keep repeating these points for all the versions of kubernetes we intend to support. minikube makes really easy to switch between them.

What is blocking me is:

  * minikube requires either kvm or VirtualBox to start the machines.
  * I don't know how this could be integrated into the salt CI pipeline.

What do you think?

**Comment by isbm** [comment]
@flavio the real life "destructive" integration tests are other story. But at least Unit tests that covering _your_ logic in the current code would be good. You know [our draconian rules](https://github.com/opensuse/salt/wiki/workflow#writing-salt-toaster-tests), right? :wink:

**Pull request merged**
